log4j-api/src/main/java/org/apache/logging/log4j/util/PropertySource.java

```diff
@@ -101,10 +101,12 @@ class Comparator implements java.util.Comparator<PropertySource>, Serializable {
         private static final long serialVersionUID = 1L;
 
         @Override
-        public int compare(final PropertySource o1, final PropertySource o2) {
-            return Integer.compare(
-                    Objects.requireNonNull(o1).getPriority(),
-                    Objects.requireNonNull(o2).getPriority());
+        public int compare(final PropertySource left, final PropertySource right) {
+            Objects.requireNonNull(left);
+            Objects.requireNonNull(right);
+            final int result = Integer.compare(left.getPriority(), right.getPriority());
+            // Two property sources can have the same priority
+            return result != 0 || left.equals(right) ? result : left.hashCode() - right.hashCode();
```

Comment:
I agree with Phillip Webb's proposal in [LOG4J2-3618](https://issues.apache.org/jira/browse/LOG4J2-3618):

> use `equals/hashCode` to determine if items can be added, and only use the Comparator for sorting

I am not able to see how this achieves that. I would be in favor of the following instead:
1. Leave this comparator as is (only compare using ranks)
2. Initialize the `PropertiesUtil.Environment#sources` using `new ConcurrentSkipListSet()`, i.e., no explicit comparator
3. Find iterations over `PropertiesUtil.Environment#sources`, and employ the comparator there (i.e., `.stream().sorted(Comparator.INSTANCE).forEach(...)`) _iff_ an order is needed.

Reply:
Fixed in 2a40dc5bfa2a368b432f057194e6241f6626bb85